src/main/java/com/google/api/gax/retrying/AbstractRetryHandler.java

```diff
@@ -0,0 +1,137 @@
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.google.api.gax.retrying;
+
+import com.google.api.gax.core.NanoClock;
+import com.google.api.gax.core.RetrySettings;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.concurrent.TimeUnit;
+import org.joda.time.Duration;
+
+/**
+ * Basic implementation of the retry {@link RetryHandler} interface. Responsible for defining and
+ * checking attempt's basic properties (execution time and count limits).
+ *
+ * This class is thread-safe, and all inheriting classes are required to be thread-safe.
+ *
+ * @param <ResponseT> response type
+ */
+public abstract class AbstractRetryHandler<ResponseT> implements RetryHandler<ResponseT> {
+
+  private final NanoClock clock;
+
+  protected AbstractRetryHandler(NanoClock clock) {
+    this.clock = clock;
+  }
+
+  /**
+   * Ensures that the retry logic hasn't exceeded neither maximum number of retries nor the total
+   * execution timeout.
+   *
+   * @param e exception thrown by the previous attempt
+   * @param nextAttemptSettings attempt settings, which will be used for the next attempt, if
+   * accepted
+   * @return {@code true} if none of the retry limits are exceeded
+   */
+  @Override
+  public boolean accept(Throwable e, RetryAttemptSettings nextAttemptSettings) {
+    RetrySettings globalSettings = nextAttemptSettings.getGlobalSettings();
+    long randRetryDelayMillis = nextAttemptSettings.getRandomizedRetryDelay().getMillis();
+    long totalTimeSpentNanos =
+        clock.nanoTime()
+            - nextAttemptSettings.getFirstAttemptStartTime()
+            + TimeUnit.NANOSECONDS.convert(randRetryDelayMillis, TimeUnit.MILLISECONDS);
+
+    long totalTimeoutMillis = globalSettings.getTotalTimeout().getMillis();
+    long totalTimeoutNanos =
+        TimeUnit.NANOSECONDS.convert(totalTimeoutMillis, TimeUnit.MILLISECONDS);
+
+    return totalTimeSpentNanos <= totalTimeoutNanos
+        && (globalSettings.getMaxAttempts() <= 0
+            || nextAttemptSettings.getAttemptCount() < globalSettings.getMaxAttempts());
+  }
```

Comment:
I tried to keep the RetryAttemptSettings class as a POJO without any logic in it (similar to the RetrySettings). It makes sense, as the attemptSettings represent the settings themselves, which can be interpreted differently, depending on context/client/type of client.